File: Backward chaining.pl
% Backward Chaining

fact(fly).
fact(cough).

rule(furry, [fly, cough]).
rule(rest, [fly]).
rule(doctor_visit, [furry, rest]).

backward_chaining(Goal) :-
    fact(Goal).

backward_chaining(Goal) :-
    rule(Goal, Conditions),
    prove_all(Conditions).

prove_all([]).

prove_all([Condition | Rest]) :-
    backward_chaining(Condition),
    prove_all(Rest).

start :-
    write('BACKWARD CHAINING'),
    nl,
    write('Enter goal: '),
    read(Goal),
    (
        backward_chaining(Goal)
        ->
        write('Goal can be proved from facts.')
        ;
        write('Goal cannot be proved from given facts.')
    ),
    nl.

File: Backward chaining.py
# Backward Chaining

facts = {
    "fly",
    "cough"
}

rules = {
    "furry": ["fly", "cough"],
    "rest": ["fly"],
    "doctor_visit": ["furry", "rest"]
}


def backward_chaining(goal, facts, rules, visited=None):

    if visited is None:
        visited = set()

    # If goal is already a fact
    if goal in facts:
        return True

    # Avoid infinite loops
    if goal in visited:
        return False

    visited.add(goal)

    # If no rule can prove the goal
    if goal not in rules:
        return False

    # Check all conditions
    for condition in rules[goal]:
        if not backward_chaining(condition, facts, rules, visited):
            return False

    return True


print("BACKWARD CHAINING")
print("------------------")

goal = input("Enter goal: ").lower()

if backward_chaining(goal, facts, rules):
    print("Goal:", goal, "can be proved from facts.")
else:
    print("Goal:", goal, "cannot be proved from given facts.")
